docs/source/amp.rst

```diff
@@ -7,31 +7,199 @@ Automatic Mixed Precision package - torch.cuda.amp
-to update the parameters, so the scale factor does not interfere with the learning rate.
+Each parameter's gradient (``.grad`` attribute) should be unscaled before the optimizer
+updates the parameters, so the scale factor does not interfere with the learning rate.
 
 .. autoclass:: GradScaler
     :members:
+
+.. _autocast-op-reference:
+
+Autocast Op Reference
+^^^^^^^^^^^^^^^^^^^^^
+
+Op Eligibility
+--------------
+Only CUDA ops are eligible for autocasting.
+
+Ops that run in ``float64`` or non-floating-point dtypes are not eligible, and will
+run in these types whether or not autocast is enabled.
+
+Only out-of-place ops and Tensor methods are eligible.
+In-place variants and calls that explicitly supply an ``out=...`` Tensor
+are allowed in autocast-enabled regions, but won't receive autocasting.
+For example, in an autocast-enabled region ``a.addmm(b, c)`` is guaranteed to run
+in ``float16``, but ``a.addmm_(b, c)`` and ``a.addmm(b, c, out=d)`` may not.
+For best performance and stability, prefer out-of-place ops in autocast-enabled
+regions.
+
+Op-Specific Behavior
+--------------------
+
+The following lists describe the behavior of eligible ops in autocast-enabled regions.
+Some are ``torch`` functions, some are ``torch.nn.functional`` functions,
+and some are :class:`torch.Tensor` methods.  Some are available in more than one of those
+three namespaces.  Some are called by the forward method of :class:`torch.nn.Module`\ s.
+Autocasting occurs in the C++ backend, so any listed op may be called however it's available
+in Python and still receive autocasting.
+
+Ops not listed below do not receive autocasting.  They run in the type
+defined by their inputs.  However, autocasting may still change the type
+in which unlisted ops run if they're downstream from autocasted ops.
+
+If an op is unlisted, we assume it's safe to run in ``float16`` without impairing
+convergence.  If you encounter an unlisted op that causes convergence problems
+in ``float16``, please file an issue.
+
+Ops that run in ``float16``
+"""""""""""""""""""""""""""
+These ops (if eligible) execute in ``float16`` and produce ``float16`` output.
+
+``__matmul__``,
+``addbmm``,
+``addmm``,
+``addmv``,
+``addr``,
+``baddbmm``,
+``bmm``,
+``chain_matmul``,
+``conv1d``,
+``conv2d``,
+``conv3d``,
+``conv_transpose1d``,
+``conv_transpose2d``,
+``conv_transpose3d``,
+``conv_tbc``,
+``linear``,
+``matmul``,
+``mm``,
+``mv``,
+``prelu``
+
+Ops that run in ``float32``
+"""""""""""""""""""""""""""
+These ops (if eligible) execute in ``float32`` and produce ``float32`` output.
+
+``__pow__``,
+``__rdiv__``,
+``__rpow__``,
+``__rtruediv__``,
+``acos``,
+``asin``,
+``binary_cross_entropy_with_logits``,
+``cosh``,
+``cosine_embedding_loss``,
+``cdist``,
+``cosine_similarity``,
+``cross_entropy``,
+``cumprod``,
+``cumsum``,
+``dist``,
+``erfinv``,
+``exp``,
+``expm1``,
+``gelu``,
+``group_norm``,
+``hinge_embedding_loss``,
+``kl_div``,
+``l1_loss``,
+``layer_norm``,
```

Comment:
Batch norm is alright with fp16 inputs as long as the weights are FP32.  Layer norm, I don't know, it's probably fine too (I think it uses batch norm for some code paths).  The initial native lists follow apex closely (with some additions like chain_matmul, the only subtractions are certain interpolate functions that were on the Apex blacklist because older Pytorches used slow FP16 atomics).  I'm not sure why Carl put layer_norm on the blacklist to begin with, but for the initial native integration, I want to minimize changes relative to Apex ensure anything convergent with apex will also converge with native.  Then lists can be refined over time without user-visible changes.